Fixing circular dependencies between NPC, action, and game-state files

bvvill opened the pull request:
Changes to the `game` module
 
- `npc/npc.h` was included twice, so I removed one of them
- included `npc/dialogue.h`, and moved the dialogue execution functions `start_conversation` and `run_conversation_step`, which have `game_t *game` as a parameter, from the `npc/dialogue` module to the `game` module
 
Changes to the `room` module
 
- removed the forward declarations of `npcs_in_room_t` and `npc_t`
- added `auto_gen_movement` and its helper functions (which have a `room_list_t` as a parameter) from the `npc_mov` module
   - also fixed the implementation of iterating through a linked list that was found in `auto_gen_movement`
- added `npc_one_move` and its helper function (which have a `room_hash_t` as a parameter) from the `rooms-npc` module
- added `transfer_all_npc_items` (which has a `room_t` as a parameter) from the `npc` module
 
Changes to the `npc/dialogue` module
 
- removed the dialogue executions functions described earlier, allowing the module's dependency of `game.h` to be removed
- included `cli/util.h` since it was removed from removing `game.h`
 
Changes to the `npc` module
 
- removed the forward declarations of `npc_battle_t`, `hostility_t`, `npc_mov_t`, and `convo_t`
 
Changes to the `npc_battle` module
 
- moved the `transfer_all_npc_items` to `room.h` as described earlier, allowing the removal of the dependency of `npc.h` and the forward declaration of `npc_t`
 
Changes to the `npc_move` module
 
- corrected a fair amount of typos
- moved the `auto_gen_movement` function to the `game` module as described earlier, allowing the removal of the dependency of `game.h` and the `game_t` struct
- defined a doubly-linked list of strings, which are meant to be `room_id`s, called `room_id_dll`, as using `room_list_t`s for every NPCs (I'm imagining a larger game than a few of the examples have) movement path would use a large amount of space, and the only information about a `room` that is ever used in this module is its `room_id`. As described in my previous comment, I chose a doubly-linked list also because of how easy it is to "reverse" a path.
- Now that `game.h` is no longer included, it was necessary to include the `utlist` and `uthash` header files
- Any function that previously relied on a `room_list_t` now relies on a `room_id_dll_t`, and any function that relied on a `room_t` now relies on a `char *room_id`, and their implementations in `npc_move.c` were appropriately updated
- As described in issue 1240 / PR1268, `unsigned int npc_path_pos`, which holds the index of the npc's position in its movement path, and `uint8_t npc_path_reversed`, which stores whether or not an NPC is moving forward (the direction of `*next` in the `room_id_dll`, has value 0) or backward (the direction of `*prev` in the `room_id_dll`, has value 1) along its movement path, have been added to the `npc_mov_t` struct in order to solve a bug when an NPCs movement path contains a particular room more than once. Functions to "get" these values from a `npc_mov_t` have also been implemented
- Three other new functions, `get_next_npc_room_id`, `move_npc_mov` (which replaces , and `delete_room_id_dll`, were also implemented, see `npc_move.h` for details
 
Changes to the `rooms-npc` module
 
- moved `npc_one_move` to the `room` module as described earlier, allowing the removal of the forward declarations of `npc_t` and `npc_hash_t`
 
Other Changes
 
- Going from `room_t` and `room_list_t` to `char *room_id` and `room_id_dll_t` in the `npc_move` had a large change amplification within many test and example files, including `test_npc_move.c`, `test_rooms_npc.c`, `test_npc.c`, `quest-example.c`, and `battle_move_example_2021.c`
- Removed the unnecessary inclusion of `npc/npc_battle.h` within `battle/battle_flow.h`, which was causing a circular dependency
- There was a function in `cli/cli_ctx.h` that was named differently compared to its implementation, and didn't cause an error until some of the circular dependencies were severed, so I fixed that rq
 
Going Forward

- There is still much work to be done regarding circular dependencies within chiventure, one that sticks out to me involves some of the `skilltrees` modules, `player.h`, `ctx.h`, `item.h`, and `game.h` of course. Even if my PR gets approved i plan to leave the issue in the backlog for other teams who want to tackle these dependencies in their respective territories.

bvvill committed:
fixing the need to foward declare anything within the entire npc folder. Some functions were moved to more relevant modules, and in npc_move the need for room_list_t was replaced by a doubly-linked-list of strings of room_id

bvvill committed:
putting the forward declarations back in game.h

bvvill committed:
some more changes

bvvill committed:
fixing some more of the dependencies I missed within the room, npc, and the rooms-npc modules

bvvill committed:
moving along one failed build at a time

bvvill committed:
chugging along

bvvill committed:
fixing type and LL syntax errors in get_num_rooms

bvvill committed:
mistakenly initialized the count integer for LL_COUNT

bvvill committed:
trying to fix npc_one_move in room.h

bvvill committed:
i was being dumb with macros again

bvvill committed:
forgot semicolon, won't fix much tho lol

bvvill committed:
having difficulty with HASH_FIND arguements

bvvill committed:
hopefully we get past 34% with this one

bvvill committed:
editing test_npc.c to relfect the changes in the npc_mov_t struct

bvvill committed:
updating test_rooms_npc.c to reflect changes in in the npc_move and rooms-npc modules

bvvill committed:
test_npc_battle.c seems to depend on a lot

bvvill committed:
now updating test_npc_move.c to reflect the changes in the npc_move, room, and other modules

bvvill committed:
now updating test_npc_move.c to reflect changes in the other modules

bvvill committed:
forgot semicolon

bvvill committed:
now updating a couple of the example files

bvvill committed:
not sure where to start

bvvill committed:
deleted a parenthesis, doesn't fix anything new tho

bvvill committed:
working on sorting out the failed npc tests

bvvill committed:
had the HASH_FIND 4th arguement all wrong

bvvill committed:
breh

bvvill committed:
breh 2

bvvill committed:
breh 3, i'll be going to bed soon i promise

bvvill committed:
breh 4

bvvill committed:
sorry for the future behemoth

bvvill committed:
addressing weird errors in test 11, not sure if they're real

bvvill committed:
okay they aren't real, i should've checked first, i'm done for now, goodnight.

bvvill committed:
moving on with purging the need for forward declarations in the game-state folder

bvvill committed:
fixing errors p1

bvvill committed:
fixing errors p2

bvvill committed:
Revert "fixing errors p2"

This reverts commit bc284fa385e825283ca8801ad703c278a6d666ce.

bvvill committed:
Revert "fixing errors p1"

This reverts commit e6f9edb9ca0acbee7caeab7993c5f90a23169364.

bvvill committed:
Revert "moving on with purging the need for forward declarations in the game-state folder"

This reverts commit e27e839049acefe62a4bc3f9616494075f485389.

bvvill committed:
Merge branch 'dev' into 1112-fix-circular-dependencies-in-npc-actions-rooms-and-game-state-files

Commit:
Merge branch 'dev' into 1112-fix-circular-dependencies-in-npc-actions-rooms-and-game-state-files

junasano commented:
1. Typo on line 638 of game-state/src/game.c file. "dialogue.h" should be "game.h" instead

bvvill committed:
Merge branch 'dev' into 1112-fix-circular-dependencies-in-npc-actions-rooms-and-game-state-files

bvvill committed:
fixing merging mistakes

bvvill committed:
forgot bracket

bvvill committed:
forgot a function

bvvill committed:
i'm dumb lol

bvvill committed:
replacing "See dialogue.h" with "See game.h" in game.c

bvvill commented:
> 1. Typo on line 638 of game-state/src/game.c file. "dialogue.h" should be "game.h" instead

thank you, it's now been fixed

Commit:
fixed some indentation inconsistencies in test_rooms_npc.c

Commit:
Merge branch '1112-fix-circular-dependencies-in-npc-actions-rooms-and-game-state-files' of github.com:uchicago-cs/chiventure into 1112-fix-circular-dependencies-in-npc-actions-rooms-and-game-state-files

junasano commented:
There were some indentation inconsistencies within test_rooms_npc.c so I went ahead and fixed it, and pushed the code to the 1112 branch.

Commit:
fixed some more indentation inconsistencies

bvvill committed:
used the astyle command for formatting .c files in the npc src directory to be allman style

bvvill committed:
astyling (to allman) any .c files that were being edited in this commit for any other reason

junasano commented:
I fixed some more indentation issues. (already pushed)
I looked through in-depth, everything except the tests and examples, and have listed some suggestions below. Tests and examples looked good initially but will take a closer look tomorrow. 

Some suggestions: 
1. Change int rc = 0; to int rc = SUCCESS; to reflect naming conventions, on line 255 of room.c
2. npc_mov_new description should say: Returns newly heap allocated npc_mov struct on success, and NULL if an error occurs. (Line 143 npc_move.h)
3. maybe remove the casting (room_t *) from line 266 of room.c since all(that I saw) other mallocs don’t seem to have the casting part in Chiventure. Same issue on line 120 of npc_move.c.
4. Need to update comment from room_list_t structs to room_id_dll_t structs on line 270 and 277 of npc_move. 
5. For the function room_id_cmp, I just thought it might be a waste to simply regard the output as 0 or non-zero, especially considering all of the other functions have defined outputs. Maybe something like 0 for match and 1 for non-match to stay a little consistent with SUCCESS and FAILURE. (npc_move.c)
6. styling issue: get rid of parenthesis for if statement starting on line 265 and 312 of npc_move.c, and for else statement starting on line 285 and 333. Since they are both just simple return statements. 
7. Technically not part of this PR but rename rooms-npc.c to npc_rooms.c to match other naming conventions (could also bring up during next meeting)

bvvill committed:
Implementing Jun's suggestions

bvvill committed:
astyling

bvvill committed:
renaming the rooms-npc module to rooms_npc

bvvill commented:
> I fixed some more indentation issues. (already pushed) I looked through in-depth, everything except the tests and examples, and have listed some suggestions below. Tests and examples looked good initially but will take a closer look tomorrow.
> 
> Some suggestions:
> 
> 1. Change int rc = 0; to int rc = SUCCESS; to reflect naming conventions, on line 255 of room.c
> 2. npc_mov_new description should say: Returns newly heap allocated npc_mov struct on success, and NULL if an error occurs. (Line 143 npc_move.h)
> 3. maybe remove the casting (room_t *) from line 266 of room.c since all(that I saw) other mallocs don’t seem to have the casting part in Chiventure. Same issue on line 120 of npc_move.c.
> 4. Need to update comment from room_list_t structs to room_id_dll_t structs on line 270 and 277 of npc_move.
> 5. For the function room_id_cmp, I just thought it might be a waste to simply regard the output as 0 or non-zero, especially considering all of the other functions have defined outputs. Maybe something like 0 for match and 1 for non-match to stay a little consistent with SUCCESS and FAILURE. (npc_move.c)
> 6. styling issue: get rid of parenthesis for if statement starting on line 265 and 312 of npc_move.c, and for else statement starting on line 285 and 333. Since they are both just simple return statements.
> 7. Technically not part of this PR but rename rooms-npc.c to npc_rooms.c to match other naming conventions (could also bring up during next meeting)

Thank you for the thorough review, Jun, everything you listed has been implemented with the exception of #5, but I changed the description of the function as it was a little misleading: it's not just a "0 or non-zero" function, it's a `strcmp()`, so the outputs have lexigraphical significance and can be used for sorting, if need be.

junasano commented:
The changes look good!
One last thing about the tests. For move_npc_definite test, maybe add more cases for where the function returns 1(when npc reaches end of path) and 3 (when npc has no where to move). The logic in the implementation seems good to the extent of my knowledge but maybe worth adding tests cases as well in case. Same thing with move_npc_indefinite. They are in the test_npc_move.c file. 
Other than that everything looks good.

bvvill committed:
Merge branch 'dev' into 1112-fix-circular-dependencies-in-npc-actions-rooms-and-game-state-files

bvvill committed:
adding additional tests for move_npc_(in)definite

bvvill committed:
that's what I get for copying and pasting

bvvill commented:
> The changes look good! One last thing about the tests. For move_npc_definite test, maybe add more cases for where the function returns 1(when npc reaches end of path) and 3 (when npc has no where to move). The logic in the implementation seems good to the extent of my knowledge but maybe worth adding tests cases as well in case. Same thing with move_npc_indefinite. They are in the test_npc_move.c file. Other than that everything looks good.

yes I'm definitely guilty of neglecting the tests, they've been implemented, thank you!

junasano commented:
The additional tests looks good and the PR is ready to go

bvvill committed:
Merge branch 'dev' into 1112-fix-circular-dependencies-in-npc-actions-rooms-and-game-state-files

junasano reviewed (approved):
Everything looks good and the suggestions I made were incorporated. PR #1286 is ready to go.

lilyehsani commented on src/npc/src/npc_move.c:
```diff
@@ -158,6 +153,44 @@ char* get_npc_curr_room_id(npc_mov_t *npc_mov)
+{
+    room_id_dll_t *current_room;
+
+    if (npc_mov->mov_type == NPC_MOV_INDEFINITE)
```
Change to Allman Style brackets for consistency.

lilyehsani commented on src/npc/src/npc_move.c:
```diff
@@ -158,6 +153,44 @@ char* get_npc_curr_room_id(npc_mov_t *npc_mov)
+
+    if (direction == NPC_MOV_ORIGINAL)
+    {
+        if (current_room->next == NULL) return NULL;
```
Change to Allman Style brackets for consistency.

lilyehsani commented on src/npc/src/npc_move.c:
```diff
@@ -174,260 +207,157 @@ unsigned int get_npc_path_direction(npc_mov_t *npc_mov)
-        count++;
-    }
-    return count;
+    if (mov_type == NPC_MOV_DEFINITE)
```
Change to Allman Style brackets for consistency.

lilyehsani commented on src/quests/examples/quest-example.c:
```diff
@@ -311,6 +311,16 @@
 //     reward_t *reward_passive = reward_new(0, item_new("Portal Gun", "this gun can create portals on special walls",
 //     "Reward for completing passive missions."));
 
+    /*quest layout: start in room1 -> go to room2 -> go to room3 -> get emerald -> go to room4 -> fight wolf and WIN -> get potion -> meet npc for reward
```
Looks like this showed up in a merge resolution by accident because it repeats what is directly below. Please remove it.

lilyehsani replied:
(Resolved before review was posted)

bvvill committed:
fixing merge conflict duplicate in quest-example

lilyehsani reviewed (changes requested):
I am really impressed with the work done here! Jun also completed a very thorough review, so not much from me, just a few style comments. Great work!

bvvill committed:
fixing incorrect styling

bvvill committed:
Merge branch 'dev' into 1112-fix-circular-dependencies-in-npc-actions-rooms-and-game-state-files

bvvill committed:
hopefully fixing an implicit declaration

bvvill committed:
Merge branch 'dev' into 1112-fix-circular-dependencies-in-npc-actions-rooms-and-game-state-files

bvvill commented:
> I am really impressed with the work done here! Jun also completed a very thorough review, so not much from me, just a few style comments. Great work!

Thank you! The style corrections have been committed.

lilyehsani reviewed (approved):
LGTM!

bvvill commented:
Closing Statement:

Changes to the `game` module 
* `npc/npc.h` was included twice, so I removed one of them
* included `npc/dialogue.h`, and moved the dialogue execution functions `start_conversation` and `run_conversation_step`, which have `game_t *game` as a parameter, from the `npc/dialogue` module to the `game` module 

Changes to the `room` module 
* removed the forward declarations of `npcs_in_room_t` and `npc_t`
* added `auto_gen_movement` and its helper functions (which have a `room_list_t` as a parameter) from the `npc_mov` module   
   * also fixed the implementation of iterating through a linked list that was found in `auto_gen_movement`
* added `npc_one_move` and its helper function (which have a `room_hash_t` as a parameter) from the `rooms-npc` module
* added `transfer_all_npc_items` (which has a `room_t` as a parameter) from the `npc` module

Changes to the `npc/dialogue` module
> 
* removed the dialogue executions functions described earlier, allowing the module's dependency of `game.h` to be removed
* included `cli/util.h` since it was removed from removing `game.h`
 
Changes to the `npc` module

* removed the forward declarations of `npc_battle_t`, `hostility_t`, `npc_mov_t`, and `convo_t`
* renamed `rooms-npc.h/.c` to `rooms_npc.h/.c` to align with chiventure naming conventions

Changes to the `npc_battle` module

* moved the `transfer_all_npc_items` to `room.h` as described earlier, allowing the removal of the dependency of `npc.h` and the forward declaration of `npc_t`

Changes to the `npc_move` module

* corrected a fair amount of typos
* moved the `auto_gen_movement` function to the `game` module as described earlier, allowing the removal of the dependency of `game.h` and the `game_t` struct
* defined a doubly-linked list of strings, which are meant to be `room_id`s, called `room_id_dll`, as using `room_list_t`s for every NPCs (I'm imagining a larger game than a few of the examples have) movement path would use a large amount of space, and the only information about a `room` that is ever used in this module is its `room_id`. As described in my previous comment, I chose a doubly-linked list also because of how easy it is to "reverse" a path.
* Now that `game.h` is no longer included, it was necessary to include the `utlist` and `uthash` header files
* Any function that previously relied on a `room_list_t` now relies on a `room_id_dll_t`, and any function that relied on a `room_t` now relies on a `char *room_id`, and their implementations in `npc_move.c` were appropriately updated
* As described in issue 1240 / PR1268, `unsigned int npc_path_pos`, which holds the index of the npc's position in its movement path, and `uint8_t npc_path_reversed`, which stores whether or not an NPC is moving forward (the direction of `*next` in the `room_id_dll`, has value 0) or backward (the direction of `*prev` in the `room_id_dll`, has value 1) along its movement path, have been added to the `npc_mov_t` struct in order to solve a bug when an NPCs movement path contains a particular room more than once. Functions to "get" these values from a `npc_mov_t` have also been implemented
* Three other new functions, `get_next_npc_room_id`, `move_npc_mov` (which replaces , and `delete_room_id_dll`, were also implemented, see `npc_move.h` for details

Changes to the `rooms-npc` module

* moved `npc_one_move` to the `room` module as described earlier, allowing the removal of the forward declarations of `npc_t` and `npc_hash_t`

Other Changes

* Going from `room_t` and `room_list_t` to `char *room_id` and `room_id_dll_t` in the `npc_move` had a large change amplification within many test and example files, including `test_npc_move.c`, `test_rooms_npc.c`, `test_npc.c`, `quest-example.c`, and `battle_move_example_2021.c`
* Removed the unnecessary inclusion of `npc/npc_battle.h` within `battle/battle_flow.h`, which was causing a circular dependency
* There was a function in `cli/cli_ctx.h` that was named differently compared to its implementation, and didn't cause an error until some of the circular dependencies were severed, so I fixed that rq

Going Forward

* There is still much work to be done regarding circular dependencies within chiventure, one that sticks out to me involves some of the `skilltrees` modules, `player.h`, `ctx.h`, `item.h`, and `game.h` of course.

lilyehsani commented:
Issue Score: Excellent

Comments:
Great work on this PR! The summary is very thorough.

<!-- {"score": "excellent", "minor": false} -->
